pkg/controllers/management/nodepool/nodepool.go

```diff
@@ -56,15 +62,73 @@ func (c *Controller) Create(nodePool *v3.NodePool) (runtime.Object, error) {
 	return nodePool, nil
 }
 
+func (c *Controller) needsReconcile(nodePool *v3.NodePool) bool {
+	changed, _, err := c.createOrCheckNodes(nodePool, true)
+	if err != nil {
+		logrus.Debugf("[nodepoool] error checking pool for reconciliation: %s", err)
+	}
+
+	return changed
+}
+
+func (c *Controller) reconcile(nodePool *v3.NodePool) {
+	_, qty, err := c.createOrCheckNodes(nodePool, false)
+	if err != nil {
+		logrus.Errorf("[nodepool] reconcile erorr, create or check nodes: %s", err)
+	}
+
+	var np *v3.NodePool
+	if qty != nodePool.Spec.Quantity {
+		// pool was scaled down, need to update qty
+		np = nodePool.DeepCopy()
+		np.Spec.Quantity = qty
+	} else {
+		np = nodePool
+	}
+
+	np.Annotations[ReconcileAnnotation] = "updated"
+	if _, err := c.NodePools.Update(np); err != nil {
+		logrus.Errorf("[nodepool] error updating nodepool after reconcile: %s", err)
+	}
+}
+
 func (c *Controller) Updated(nodePool *v3.NodePool) (runtime.Object, error) {
 	obj, err := v32.NodePoolConditionUpdated.Do(nodePool, func() (runtime.Object, error) {
-		return nodePool, c.reconcile(nodePool)
+		anno, _ := nodePool.Annotations[ReconcileAnnotation]
+		if anno == "" {
+			if c.needsReconcile(nodePool) {
+				logrus.Debugf("[nodepool] reconcile needed for %s", nodePool.Name)
+				np, err := c.setReconcileAnnotation(nodePool, "updating")
+				if err != nil {
+					return nodePool, err
+				}
+				go c.reconcile(np)
+				return nil, nil
+			}
+		} else if anno == "updated" {
+			nodePool.Annotations[ReconcileAnnotation] = ""
+			return c.NodePools.Update(nodePool)
+		} else if strings.HasPrefix(anno, "updating/") {
+			// gate updating the node pool to every 20s
+			pieces := strings.Split(anno, "/")
+			t, err := time.Parse(time.RFC3339, pieces[1])
+			if err != nil || int(time.Since(t)/time.Second) > 20 {
+				nodePool.Annotations[ReconcileAnnotation] = ""
```

Comment:
is the main purpose of this update to call the reconcile check again? 
Since this func only gets called on an update sync, I'm confused why we need to clean up the annotation every 20s vs checking this on L98 itself: 

`if anno == "" || timeToGateUpdate`

Reply:
the annotation has a timestamp set to it. if something failed (server restarted or anything) and the go routine bombed we need to clear the annotation and start again. it's just safety to give things another try if they failed. similar code is in the cluster provisioned controller